src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs

```diff
@@ -766,5 +767,45 @@ private async Task<bool> CompleteOutstandingActiveRequests(string targetName)
 
             return false;
         }
+
+        private void ComputeAfterTargetFailures(string[] targetNames)
+        {
+            // Cache intermediate results so we don't visit a target more than once.
+            Dictionary<string, bool> resultsWithAfterTargets = new Dictionary<string, bool>();
+
+            foreach (string targetName in targetNames)
+            {
+                if (_buildResult.ResultsByTarget.ContainsKey(targetName))
+                {
+                    // Queue of targets waiting to be processed, seeded with the specific target for which we're computing AfterTargetsHaveFailed.
+                    Queue<string> targetsToCheckForAfterTargets = new Queue<string>();
+                    targetsToCheckForAfterTargets.Enqueue(targetName);
+
+                    while (targetsToCheckForAfterTargets.Count > 0)
+                    {
+                        string targetToCheck = targetsToCheckForAfterTargets.Dequeue();
+                        IList<TargetSpecification> targetsWhichRunAfter = _requestEntry.RequestConfiguration.Project.GetTargetsWhichRunAfter(targetToCheck);
+
+                        foreach (TargetSpecification afterTarget in targetsWhichRunAfter)
+                        {
+                            if (!resultsWithAfterTargets.TryGetValue(afterTarget.TargetName, out bool afterTargetFailed))
+                            {
+                                _buildResult.ResultsByTarget.TryGetValue(afterTarget.TargetName, out TargetResult result);
+                                afterTargetFailed = result != null && result.ResultCode == TargetResultCode.Failure && !result.TargetFailureDoesntCauseBuildFailure;
+
+                                resultsWithAfterTargets[afterTarget.TargetName] = afterTargetFailed;
+                                if (afterTargetFailed)
+                                {
+                                    _buildResult.ResultsByTarget[targetName].AfterTargetsHaveFailed = true;
+                                }
+
+                                // We haven't seen this target yet, add it to the list to check.
+                                targetsToCheckForAfterTargets.Enqueue(afterTarget.TargetName);
```

Comment:
I'm probably not 100% sure about the flow, as I see a potential never ending `while`: if we enqueue items inside the body of the loop, on the same queueue we're monitoring to exit the loop...how do we protect against infinite spinning? 

Due diligence would be for me to clone the branch and step through.

Reply:
I would also not be 100% sure, despite having stepped through the code. I do have a couple of good reasons to believe that this program eventually terminates, though:

1) We are operating on an internal representation of the build files which, presumably, has validation logic in place to make sure that there are no cycles in the task graph. Handling cycles later, like in this case when actually executing the tasks, would be very problematic.

2) If you squint hard enough, an analogous loop exists in `ProcessTargetStack` in this same file.

3) Pretty much the same loop exists in `GetAfterTargetsForDefaultTargets` which I am deleting as part of this PR.

Reply:
Targets definitely cycle-detection logic. I'm not sure about tasks, but I'm also having trouble making tasks call each other, since there isn't a nice way to do that, as far as I know.

Reply:
What I'm more wondering about is how it gets the value if the aftertarget hasn't run yet. It seems to use `_buildResult.ResultsByTarget.TryGetValue` to get results, presumably from other nodes that are building the aftertargets, but does that work in a single proc build?

Reply:
Reading the logic in this file and `ProcessTargetStack` specifically, it looks like we are guaranteed that aftertargets have run by the time we get to this point, and that `_buildResult` contains their results.

For what it's worth I see it working in single proc build, both in unit tests and using the repro project file in #3345. Curious to learn more about your concern.